src/plugins/discover/public/application/main/components/layout/use_discover_histogram.ts

```diff
@@ -50,233 +41,214 @@ export const useDiscoverHistogram = ({
-    (newTopPanelHeight: number | undefined) => {
-      storage.set(HISTOGRAM_HEIGHT_KEY, newTopPanelHeight);
-      setTopPanelHeight(newTopPanelHeight);
-    },
-    [storage]
-  );
+      const { query, filters, time: timeRange } = services.data.query.getState();
 
-  /**
-   * Time interval
-   */
-
-  const onTimeIntervalChange = useCallback(
-    (newInterval: string) => {
-      stateContainer.setAppState({ interval: newInterval });
+      api.initialize({
+        services: { ...services, uiActions: getUiActions() },
+        localStorageKeyPrefix: 'discover',
+        disableAutoFetching: true,
+        initialState: {
+          dataView,
+          query,
+          filters,
+          timeRange,
+          chartHidden,
+          timeInterval,
+          breakdownField,
+          searchSessionId,
+          totalHitsStatus: totalHitsStatus.toString() as UnifiedHistogramFetchStatus,
+          totalHitsResult,
+          requestAdapter: inspectorAdapters.requests,
+        },
+      });
     },
-    [stateContainer]
+    [
+      dataView,
+      inspectorAdapters.requests,
+      isInitialized,
+      savedSearchData$.totalHits$,
+      searchSessionId,
+      services,
+      stateContainer.appState,
+    ]
   );
 
-  /**
-   * Total hits
-   */
-
-  const [localHitsContext, setLocalHitsContext] = useState<UnifiedHistogramHitsContext>();
+  const setUnifiedHistogramApi = useCallback((api: UnifiedHistogramApi | null) => {
+    if (!api) {
+      return;
+    }
 
-  const onTotalHitsChange = useCallback(
-    (status: UnifiedHistogramFetchStatus, result?: number | Error) => {
-      if (result instanceof Error) {
-        // Display the error and set totalHits$ to an error state
-        sendErrorTo(data, savedSearchData$.totalHits$)(result);
-        return;
-      }
+    setUnifiedHistogram(api);
+  }, []);
 
-      const { fetchStatus, recordRawType } = savedSearchData$.totalHits$.getValue();
+  // Initialize Unified Histogram when the first search is requested
+  useEffect(() => {
+    initializeUnifiedHistogram(unifiedHistogram);
+  }, [initializeUnifiedHistogram, isInitialized, unifiedHistogram]);
 
-      // If we have a partial result already, we don't want to update the total hits back to loading
-      if (fetchStatus === FetchStatus.PARTIAL && status === UnifiedHistogramFetchStatus.loading) {
-        return;
-      }
+  /**
+   * State syncing
+   */
 
-      // Set a local copy of the hits context to pass to unified histogram
-      setLocalHitsContext({ status, total: result });
+  useEffect(() => {
+    if (!unifiedHistogram?.initialized) {
+      return;
+    }
 
-      // Sync the totalHits$ observable with the unified histogram state
-      savedSearchData$.totalHits$.next({
-        fetchStatus: status.toString() as FetchStatus,
-        result,
-        recordRawType,
+    const subscription = unifiedHistogram
+      .getState$(({ lensRequestAdapter, chartHidden, timeInterval, breakdownField }) => ({
```

Comment:
The selector functions should be defined outside of the component.

Reply:
I switched away from this approach and created `createStateSyncObservable` and `createTotalHitsObservable` utility functions to do this work instead.